### PR TITLE
Add manual connection option

### PR DESCRIPTION
**Pull request by haslinghuis**
- Have observed users clicking the connect button without any connection available making manual connection an option would fix most `Failed to open serial port` events. During this process port_handler has been optimized.
- manual optional is now optional - as most users would only use serial connection
- removes brain overload after refactoring some code in porthandler
- made currentPorts global simplifying code
- during development encountered some issues which are fixed in:
  #3932
  #3933
  #3935

image

image

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by McGiverGim** [comment]
I don't understand. What is the problem with JQuery? I understand is old technology and if we can get rid of it, great. But remove it to go with plain javascript, does not seem a good option to me.
The idea was go to Vue (or any other technology that we decide and simplifies the code) but not to plain javascript.

**Comment by haslinghuis** [comment]
Agree Vue was adapted some time ago and thought this move was creating a path for further migration.

**Comment by haslinghuis** [comment]
@McGiverGim @chmelevskij please assist in getting this to use Vue as requested.

**Comment by nerdCopter** [comment]
Unsure about this.

If users clicking too soon, enabling the buttons could be moved to after ports being populated similarly to https://github.com/emuflight/EmuConfigurator/pull/458/files
However, this is only effective at first launch.

If port handler needs independent optimization, can it be a separate PR? or is it dependent on this new logic?

**Comment by haslinghuis** [comment]
Going to rewrite the PR so it only adds manual connection option - as far is possible as some changes were needed to make it work. Hiding manual mode would prevent connecting to a non-existent device.

**Comment by nerdCopter** [comment]
- i don't feel this is needed, but i also do not know the statistics on the offending clicks.
- it will also add another step to vtx-slayer's https://pr0p.dev/ , but such is an external issue.

**Comment by haslinghuis** [comment]
Agree but assume most users won't need manual connection as hiding this simplifies UI. That's all :)

**Comment by haslinghuis** [comment]
Put a lot of work in this - required lot's of testing different scenario's - but it has some other fixes too:

image

```
2024-01-04 @17:31:10 -- Serial port successfully closed
2024-01-04 @17:31:10 -- Device - Rebooting
2024-01-04 @17:31:10 -- Using cached builds information for https://build.betaflight.com/api/targets.
2024-01-04 @17:31:10 -- Using cached builds information for https://build.betaflight.com/api/targets.
2024-01-04 @17:31:10 -- Using cached builds information for https://build.betaflight.com/api/configurator/sponsors/dark/landing.
2024-01-04 @17:31:10 -- Using cached builds information for https://build.betaflight.com/api/configurator/sponsors/dark/flash.
2024-01-04 @17:31:11 -- Using cached builds information for https://build.betaflight.com/api/targets.
2024-01-04 @17:31:11 -- Using cached builds information for https://build.betaflight.com/api/configurator/sponsors/dark/flash.
2024-01-04 @17:31:11 -- Using cached builds information for https://build.betaflight.com/api/targets.
2024-01-04 @17:31:15 -- Query board information to preselect right firmware
2024-01-04 @17:31:15 -- MultiWii API version: 1.46.0
2024-01-04 @17:31:15 -- Using cached builds information for https://build.betaflight.com/api/targets/FLYWOOF722PRO.
2024-01-04 @17:31:15 -- Using cached builds information for https://build.betaflight.com/api/builds/4.5.0-zulu/FLYWOOF722PRO.
2024-01-04 @17:31:15 -- Using cached builds information for https://build.betaflight.com/api/releases/4.5.0-zulu/commits.
2024-01-04 @17:31:15 -- Configurator has successfully detected and verified the board: FLYWOOF722PRO
2024-01-04 @17:31:21 -- Failed to open serial port
```

This PR:

```
2024-01-04 @18:33:21 -- OS: Linux
2024-01-04 @18:33:21 -- Configurator: 10.10.0-debug-27b1c354
2024-01-04 @18:33:24 -- Using cached builds information for https://build.betaflight.com/api/targets.
2024-01-04 @18:33:25 -- Query board information to preselect right firmware
2024-01-04 @18:33:25 -- MultiWii API version: 1.46.0
2024-01-04 @18:33:25 -- Configurator has successfully detected and verified the board: FLYWOOF722PRO
2024-01-04 @18:33:28 -- Success: https://build.betaflight.com/api/builds/740b520792e3ae53af940c9ab98a7633/status
2024-01-04 @18:33:29 -- Success: /api/builds/740b520792e3ae53af940c9ab98a7633/hex
2024-01-04 @18:33:34 -- MultiWii API version: 1.46.0
2024-01-04 @18:33:34 -- Device - Rebooting to ROM BOOTLOADER
2024-01-04 @18:33:35 -- USB device successfully opened with ID: 1
2024-01-04 @18:33:35 -- Detected device with total flash size 512 KiB
2024-01-04 @18:33:35 -- Chrome API Error: Transfer stalled..
2024-01-04 @18:33:35 -- Read protection not active
2024-01-04 @18:33:43 -- Erased 512 kB of flash successfully
2024-01-04 @18:34:08 -- USB device successfully closed
2024-01-04 @18:34:11 -- Query board information to preselect right firmware
2024-01-04 @18:34:11 -- MultiWii API version: 1.46.0
2024-01-04 @18:34:11 -- Configurator has successfully detected and verified the board: FLYWOOF722PRO
```

**Comment by nerdCopter** [comment]
request real world testing from multiple devs before merge.

**Comment by ctzsnooze** [comment]
I'm all for improving port detection, but do not agree with changing to manual configuration if the user keeps pressing the 'connect' button.

On my mac, after some time of normal operation, a serial port stops opening  when I connect a flight controller.  I have looked into it, and it's because at system level, the Mac isn't seeing any connection for Betaflight to utilise. Switching to manual configuration - manually - has not been helpful on a single instance.  A reboot will fix it.  So I see no value in changing to manual configuration automatically.  The user should decide to do that.

I've not had a single instance of a port being available that wasn't automatically loaded, or at least I didn't notice it.

**Review by blckmn — DISMISSED** [review]
@haslinghuis - Can you please split this PR into bug -fixes only for 4.5.0-RC*, and then the additional feature for 4.6 or higher.

**Comment by haslinghuis** [comment]
@blckmn the issue is intermittent and non-breaking so skipping to next release.

**Comment by HThuren** [comment]
@haslinghuis can we use bluetooth with this PR ?

**Comment by haslinghuis** [comment]
In theory it should already be possible using serial mode but did not look at it.

**Review by blckmn — APPROVED** [review]

**Comment by haslinghuis** [comment]
Testing. Just a minor display issue. Need to hide that while rebooting.

image

**Comment by nerdCopter** [comment]
`3785b9f5`
`rm ~/.config/betaflight-blackbox-explorer/ betaflight-configurator/`
- clicking connect with no FC, does nothing; no errors, logs, nor console. :rocket: 
- toggling `Enable manual connection mode` properly add/removes `manual Selection` from drop-down and also shows/hides manual port field. :rocket: 
- pseudo-setup unsolderd FC worked fine.
- moving from VTX tab to LED Strip tab, went to click on an item, and tab jumped back to VTX tab.  most-likly unrelated to this PR and most likely a timing/promise issue.

**Review by nerdCopter — APPROVED** [review]

**Pull request merged**
